db/pqdb.go

```diff
@@ -0,0 +1,278 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package db
+
+import (
+	"database/sql"
+	"errors"
+	"fmt"
+
+	"github.com/lib/pq"
+
+	_ "github.com/lib/pq"
+	"github.com/servian/TechChallengeApp/model"
+)
+
+type Pqdb struct {
+}
+
+func (Pqdb) getDbInfo(cfg Config) string {
+	return fmt.Sprintf("host=%s port=%s user=%s password=%s dbname=%s sslmode=disable",
+		cfg.DbHost, cfg.DbPort, cfg.DbUser, cfg.DbPassword, cfg.DbName)
+}
+
+// RebuildDb drops the database and recreates it
+func (Pqdb) RebuildDb(cfg Config) error {
+	dbinfo := fmt.Sprintf("host=%s port=%s user=%s password=%s dbname=postgres sslmode=disable",
+		cfg.DbHost, cfg.DbPort, cfg.DbUser, cfg.DbPassword)
+
+	db, err := sql.Open("postgres", dbinfo)
+
+	if err != nil {
+		return err
+	}
+
+	defer db.Close()
+
+	query := "DROP DATABASE IF EXISTS " + cfg.DbName
+
+	fmt.Println(query)
+
+	_, err = db.Query(query)
+
+	if err != nil {
+		return err
+	}
+
+	query = fmt.Sprintf(`CREATE DATABASE %s
+WITH
+OWNER = %s
+ENCODING = 'UTF8'
+LC_COLLATE = 'en_US.utf8'
+LC_CTYPE = 'en_US.utf8'
+TABLESPACE = pg_default
+CONNECTION LIMIT = -1
+TEMPLATE template0;`, cfg.DbName, cfg.DbUser)
+
+	fmt.Println(query)
+
+	_, err = db.Query(query)
+
+	return err
+}
+
+func (p Pqdb) CreateTable(cfg Config) error {
+	dbinfo := p.getDbInfo(cfg)
+
+	db, err := sql.Open("postgres", dbinfo)
+
+	if err != nil {
+		return err
+	}
+
+	defer db.Close()
+
+	tx, err := db.Begin()
+
+	if err != nil {
+		return err
+	}
+
+	defer tx.Rollback()
+
+	query := "DROP TABLE IF EXISTS tasks CASCADE"
+
+	fmt.Println(query)
+
+	_, err = tx.Exec(query)
+
+	if err != nil {
+		return err
+	}
+
+	query = "CREATE TABLE tasks ( id SERIAL PRIMARY KEY, completed boolean NOT NULL, priority integer NOT NULL, title text NOT NULL)"
+
+	fmt.Println(query)
+
+	_, err = tx.Exec(query)
+
+	if err != nil {
+		return err
+	}
+
+	if err != nil {
+		return err
+	}
+
+	err = tx.Commit()
+
+	return err
+}
+
+func (p Pqdb) SeedData(cfg Config) error {
+
+	dbinfo := p.getDbInfo(cfg)
+
+	db, err := sql.Open("postgres", dbinfo)
+
+	if err != nil {
+		return err
+	}
+
+	defer db.Close()
+
+	tx, err := db.Begin()
+
+	if err != nil {
+		return err
+	}
+
+	defer tx.Rollback()
+
+	stmt, err := tx.Prepare(pq.CopyIn("tasks", "completed", "priority", "title"))
+
+	if err != nil {
+		return err
+	}
+
+	tasks := p.getSeedTasks()
+
+	for _, task := range tasks {
+		_, err = stmt.Exec(task.Complete, task.Priority, task.Title)
+		if err != nil {
+			return err
+		}
+	}
+
+	_, err = stmt.Exec()
+
+	if err != nil {
+		return err
+	}
+
+	err = stmt.Close()
+
+	if err != nil {
+		return err
+	}
+
+	err = tx.Commit()
+
+	return err
+}
+
+func (Pqdb) getSeedTasks() []model.Task {
```

Comment:
this is common to both implementations, so should used across both?

Reply:
moved seed data upstream.